File: groupAnagrams.py
#Leetcode Problem: 49
#key is sorted string and words are mapped to sorted string(key)
class Solution:
    def groupAnagrams(self, strs):
        anagrams=defaultdict(list)
        for i in strs:
            anagrams[tuple(sorted(i))].append(i)
        result=[]
        for k,v in anagrams.items():
            result.append(v)
        return result  

#TC: O(n klog k)  where n is total number of strings in list and k is avg length of a string
#SC: O(n)

#We can solve this by using prime numbers as well i.e. by assigning prime numbers to each letter and doing multiplication
#and putting it as key. this will take O(n*k)

File: GroupAnagrams.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// Time Complexity : O(nk) where n is the number of words each of length k
// Space Complexity : O(n) if all the words are unique.
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs == null || strs.length == 0) {
            return new ArrayList<>();
        }

        HashMap<Double, List<String>> map = new HashMap<>();

        for (int i = 0; i < strs.length; i++) {
            String str = strs[i];
            double primeproduct = calcprimeprod(str);
            if (!map.containsKey(primeproduct)) {
                map.put(primeproduct, new ArrayList<>());
            }
            map.get(primeproduct).add(str);
        }
        return new ArrayList<>(map.values());
    }

    private double calcprimeprod(String s) {
        int[] primes = { 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 71, 73, 79, 83, 89, 97,
                101, 103 };
        double product = 1;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            product = product * primes[c - 'a'];
        }
        return product;

    }

}